Add more file triggers to handle more aspects of systemd

Conan-Kudo opened the pull request:
For quite a while now, there have been file triggers to handle automatically setting up service units in upstream systemd. However, most of the actions being done by these macros upon files can be set up as RPM file triggers.

In fact, in Mageia, we had been doing this for most of these. In particular, we have file triggers in place for sysusers, tmpfiles, hwdb, and the journal.

This change adds Lua versions of the original file triggers used in Mageia, based on the existing Lua-based file triggers for service units.

In addition, we can also have useful file triggers for udev rules, sysctl directives, and binfmt directives. These are based on the other existing file triggers.

Conan-Kudo commented:
@keszybz, @ignatenkobrain: Could you please review this, as this would be highly beneficial for Fedora, too.

ignatenkobrain commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
+if posix.access("/run/systemd/system") then
+    pid = posix.fork()
+    if pid == 0 then
+        assert(posix.exec("%{_bindir}/systemd-sysusers"))
```
any reason to not use `os.execute()`?

Conan-Kudo replied:
I was following what the other triggers in this file do.

keszybz replied:
There was some good reason for that, I don't remember though.

ignatenkobrain commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
     end
 end
+
+%transfiletriggerin -P 100700 -p <lua> -- @sysusersdir@
```
please leave sysusers out for now, we need to sort it out in a different way, probably.

Conan-Kudo replied:
Why? We've been using this for the entire time Mageia has been using systemd (since Mageia 2).

It works just fine. The only thing I can think of is that maybe it should be `%filetriggerin` instead of `%transfiletriggerin`.

ignatenkobrain commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
+    end
+end
+
+%transfiletriggerin -- @udevrulesdir@
```
this should not be trigger I think, because systemd already does reloading itself

Conan-Kudo replied:
It doesn't in _all_ cases, if I remember correctly. Specifically, it won't if the rules reference files/binaries outside of the rules directory.

keszybz commented:
Thanks for working on this. Filetriggers is something we certainly want to do.

Wrt. to systemd-sysusers and systemd-tmpfiles, the plan is to extend their functionality a bit. #8058 adds some more configurability, but most importantly, it adds a mechanism that allows the systemd-sysusers invocation in scriptlets to obey local admin configuration. Current macros didn't do that. This wasn't a regression compared to the old manual `getent` and `useradd` invocations, because those didn't allow admin overrides either, but we can do better with sysusers/tmpfiles, and we should.

Unfortunately we cannot just neuter %sysusers_create, because it might be used by packages in %pre, and then the package might carry files owned by that user. For backwards compatibility, once the filetriggers or transfiletriggers are in place, we can recommend that packages which don't need to (those which don't have any files owned by created users), can drop the scriptlet and rely on the filetrigger. It might turn out that packages which need to keep it are few or even nonexistent in Fedora, but even then, we need to keep the scriptlet for backwards compatibility.

Similar considerations apply to other macros: for example `%binfmt_apply` might be used by a package right before calling a binary that uses binfmt. This is rather far-fetched, but it's hard to say what non-distro packages might be doing. At least, if we decide to neuter those macros, we should have a deprecation period, where in the next release we add a note in NEWS and in the macros file, and then in the next release we finally make it `%nil`. This is a bit more work, but makes the whole process smoother.

I'll be working updating tmpfiles similarly to sysusers in #8058, and I think this PR should wait until that's merged, so that we can have a coherent story. I think you should also drop the neutering parts of this PR, at least for now.

Conan-Kudo commented:
@keszybz I dropped the neutering of macros that took parameters originally (so sysusers, tmpfiles, and binfmt).

I've kept the neutering of the macros for udev rules, udev hwdb, and the journal catalog, as those didn't take parameters.

I've also rebased onto current master.

keszybz commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
+end
+
+%transfiletriggerin -P 100500 --  @tmpfilesdir@
+-- This script will process files installed in tmpfiles.d to
```
I think this should be wrapped to \<80 columns, looks neater.

keszybz replied:
(it's ~60 now)

keszybz commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
+
+%transfiletriggerin --  @udevhwdbdir@
+-- This script will automatically invoke hwdb update if files
+-- have been installed or updated in hwdb.d.
```
`s/hwdb.d/@udevhwdbdir@/` ?

keszybz commented on src/core/triggers.systemd.in:
```diff
@@ -69,3 +70,90 @@ if posix.access("%{_localstatedir}/lib/rpm-state/systemd/needs-reload") then
+
+%transfiletriggerin --  @catalogdir@
+-- This script will automatically invoke journal catalog update if files
+-- have been installed or updated in the catalog.
```
Here too, I think this should be explicit that only `@catalogdir@` is monitored.

Conan-Kudo commented:
@keszybz These should be addressed now.

cgwalters commented:
These are all in Lua which [conflicts with rpm-ostree's sandboxing](https://github.com/projectatomic/rpm-ostree/issues/749).  Any objections if I do a PR to rewrite in shell?

yuwata commented:
Yeah. It seems broken...
```
Yu Watanabe:
Running scriptlet: systemd-238-1.1.git738ab75.fc27.x86_64                                                                                                                                                                             12/12 
/var/tmp/rpm-tmp.OuQtV4: line 1: --: command not found
/var/tmp/rpm-tmp.OuQtV4: line 2: --: command not found
/var/tmp/rpm-tmp.OuQtV4: line 3: syntax error near unexpected token `"/run/systemd/system"'
/var/tmp/rpm-tmp.OuQtV4: line 3: `if posix.access("/run/systemd/system") then'
warning: %triggerin(systemd-238-1.1.git738ab75.fc27.x86_64) scriptlet failed, exit status 2
Non-fatal <unknown> scriptlet failure in rpm package systemd
Non-fatal <unknown> scriptlet failure in rpm package systemd
/var/tmp/rpm-tmp.ujQQgu: line 1: --: command not found
/var/tmp/rpm-tmp.ujQQgu: line 2: --: command not found
/var/tmp/rpm-tmp.ujQQgu: line 3: syntax error near unexpected token `"/run/systemd/system"'
/var/tmp/rpm-tmp.ujQQgu: line 3: `if posix.access("/run/systemd/system") then'
warning: %triggerin(systemd-238-1.1.git738ab75.fc27.x86_64) scriptlet failed, exit status 2
Non-fatal <unknown> scriptlet failure in rpm package systemd
Non-fatal <unknown> scriptlet failure in rpm package systemd
/var/tmp/rpm-tmp.o3sOET: line 1: --: command not found
/var/tmp/rpm-tmp.o3sOET: line 2: --: command not found
/var/tmp/rpm-tmp.o3sOET: line 3: syntax error near unexpected token `"/run/systemd/system"'
/var/tmp/rpm-tmp.o3sOET: line 3: `if posix.access("/run/systemd/system") then'
warning: %triggerin(systemd-238-1.1.git738ab75.fc27.x86_64) scriptlet failed, exit status 2
Non-fatal <unknown> scriptlet failure in rpm package systemd
Non-fatal <unknown> scriptlet failure in rpm package systemd
/var/tmp/rpm-tmp.UJJ95i: line 1: --: command not found
/var/tmp/rpm-tmp.UJJ95i: line 2: --: command not found
/var/tmp/rpm-tmp.UJJ95i: line 3: syntax error near unexpected token `"/run/systemd/system"'
/var/tmp/rpm-tmp.UJJ95i: line 3: `if posix.access("/run/systemd/system") then'
warning: %triggerin(systemd-238-1.1.git738ab75.fc27.x86_64) scriptlet failed, exit status 2
Non-fatal <unknown> scriptlet failure in rpm package systemd
Non-fatal <unknown> scriptlet failure in rpm package systemd
/var/tmp/rpm-tmp.khrWOI: line 1: --: command not found
/var/tmp/rpm-tmp.khrWOI: line 2: --: command not found
/var/tmp/rpm-tmp.khrWOI: line 3: --: command not found
/var/tmp/rpm-tmp.khrWOI: line 4: syntax error near unexpected token `"/run/systemd/system"'
/var/tmp/rpm-tmp.khrWOI: line 4: `if posix.access("/run/systemd/system") then'
warning: %triggerin(systemd-238-1.1.git738ab75.fc27.x86_64) scriptlet failed, exit status 2
Non-fatal <unknown> scriptlet failure in rpm package systemd
Non-fatal <unknown> scriptlet failure in rpm package systemd
```

yuwata commented:
Ah, `--` is used for comment instead of `#`.

poettering commented:
> These are all in Lua which conflicts with rpm-ostree's sandboxing. Any objections if I do a PR to rewrite in shell?

Hmm, @keszybz can you comment on that?

There are reason to use lua, where shell isn't good enough, for example when the RPM moves things around (like we did for the usrmerge) or when the shell itself is updated. However, I don't think this applies here... Hence, yeah, doing this in shell sounds fine by me, but I think @keszybz might have a better idea about RPM there...

Conan-Kudo commented:
@cgwalters The original ones were in shell, but I rewrote them to Lua to avoid all kinds of idiosyncrasies with shell. And as @poettering mentioned, having the shell update on you can be... problematic.

In Mageia, we were fine with it being in shell because urpmi actually splits transactions into internally consistent subgroups, so that the ordering was sane and you could somewhat reliably guarantee you could use a working shell. As [Mageia is moving to DNF](https://wiki.mageia.org/en/Feature:Switch_to_DNF), I have to deal with the fact that this feature [does not yet exist in DNF](https://bugzilla.redhat.com/show_bug.cgi?id=1032541). And for Red Hat/Fedora, it won't exist in Yum. SUSE folks can do this with Zypper, but it is a non-default option, and doesn't quite work exactly the same way the urpmi one does.

While it may not matter as much with them being post-transaction scriptlets, I'm a lot less comfortable with a package like systemd having that requirement, since it has the potential to trigger an enormous dependency loop. I just had to deal with something like this recently, and I don't want to deal with it again.

(Rewriting several scriptlets from Shell to Lua allowed us to drop a massive ~140 package dependency loop to a smaller chain of ~7).

For whatever it's worth, I don't want this rewritten out of Lua into shell for the reasons above.

cgwalters commented:
> While it may not matter as much with them being post-transaction scriptlets, 

Right exactly, it doesn't matter for posttrans.

Conan-Kudo commented:
@cgwalters It still triggers an enormous dependency loop, especially for chroot/container setups.

keszybz commented:
Those scripts wouldn't generate a *huge* dependency loop, because all that would be required would be shell builtins and our own binaries. At least in Fedora we already require the shell for %post and other scriptlets, so adding a dependency on the shell wouldn't change much.

This patch only adds some more lua scriptlets, we already had a bunch. So if this is broken for rpm-ostree, wasn't it already broken before?

Generally I prefer lua scriptlets to shell. They help us reduce the dependency tree, are faster, and the language is nicer. Shouldn't rpm-ostree just learn to support lua scriptlets? They are used pretty widely, and rewriting them all in shell seems like a step backwards.

Conan-Kudo commented:
It didn't in Fedora, but I had to deal with this problem in a couple of other distributions.

Also, one of the distributions I work on (OpenMandriva) uses dash instead of bash for `/bin/sh`, and I just didn't want to deal with that mess.

cgwalters commented:
Does anyone using RPM to install things *not* have `/bin/sh` in their final tree?  I would actually be *very* impressed if systemd itself was usable at runtime without `/bin/sh` at least for host contexts...some of the builtin udev rules at least use `/bin/sh` right?

So I think it'd make sense for RPM to omit `%posttrans` `/bin/sh` from `Requires`. 

But it's still not clear to me that this "dependency loop" is *actually* a problem.  I mean again you keep talking about the loop, but what was the actual problem?

> Shouldn't rpm-ostree just learn to support lua scriptlets?

It's linked in the above issue but TL;DR: doing so would *directly* conflict with our script sandboxing.  See for example: https://github.com/projectatomic/rpm-ostree/pull/888
Personally, I think that's pretty damn cool.  It has clear value and solves real-world problems; it's essential for our "packages installed offline" story.  Running Lua in-process totally defeats that.  In theory we could do out-of-process but then the speed argument goes entirely away (if there was really ever a speed issue which I seriously doubt). 

Hmm, now that I think about it more...these rules don't handle rpm-ostree style offline updates.  For us `posix.access("/run/systemd/system")` will always be False.

Concretely for example, when rpm-ostree is preparing a new update, we *don't* want the Lua rules to run `systemctl daemon-reload` for the current systemd!  But we do want anything that regenerates "index"es on disk, which I believe includes e.g. `systemd-hwdb update`.

Conan-Kudo commented:
> Running Lua in-process totally defeats that. In theory we could do out-of-process but then the speed argument goes entirely away (if there was really ever a speed issue which I seriously doubt).

I have no problems with the out-of-process idea. Again, I use Lua scriptlets because I like them for being cleaner and simpler to use on a wider variety of environments.

cgwalters commented:
PR in https://github.com/systemd/systemd/pull/8550

filbranden commented:
Somewhat of a tangent, but...

```lua
    pid = posix.fork()
    if pid == 0 then
        assert(posix.exec("%{_bindir}/systemd-sysusers"))
    elseif pid > 0 then
        posix.wait(pid)
    end
```

Really? Lua really doesn't have a better interface to just spawning an external command without a shell? Such as, for example, [Python's subprocess](https://docs.python.org/2/library/subprocess.html), which they got right in version 2.4?

Not to mention that it looks like whenever fork() fails, there's no handling of that error, it seems to be silently ignored...
